client/client.go

```diff
@@ -0,0 +1,389 @@
+package client
+
+import (
+	"context"
+
+	"github.com/topolvm/topolvm"
+	topolvmlegacyv1 "github.com/topolvm/topolvm/api/legacy/v1"
+	topolvmv1 "github.com/topolvm/topolvm/api/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	kind     = "LogicalVolume"
+	kindList = "LogicalVolumeList"
+)
+
+var (
+	group = topolvmv1.GroupVersion.Group
+)
+
+type wrappedReader struct {
+	client client.Reader
+	scheme *runtime.Scheme
+}
+
+var _ client.Reader = &wrappedReader{}
+
+func NewWrappedReader(c client.Reader, s *runtime.Scheme) client.Reader {
+	return &wrappedReader{
+		client: c,
+		scheme: s,
+	}
+}
+
+func (c *wrappedReader) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	switch o := obj.(type) {
+	case *unstructured.Unstructured:
+		if gvk.Group == group && gvk.Kind == kind && topolvm.UseLegacy() {
+			o.SetGroupVersionKind(topolvmlegacyv1.GroupVersion.WithKind(kind))
+			err := c.client.Get(ctx, key, o)
+			o.SetGroupVersionKind(topolvmv1.GroupVersion.WithKind(kind))
+			return err
+		}
+		return c.client.Get(ctx, key, obj)
+	case *metav1.PartialObjectMetadata:
```

Comment:
By the way, what is this `*metav1.PartialObjectMetadata` used for? Isn't it enough to handle only untyped (`*unstructured.Unstructured`) and typed (`*topolvmv1.LogicalVolume`) cases?

Reply:
>  what is this *metav1.PartialObjectMetadata used for?

According to the client-go comment, it seems to be a client for acquiring only metadata (≈ ObjectMeta).

https://github.com/kubernetes/client-go/blob/v0.24.2/metadata/metadata.go#L50-L55

However, I personally believe that there are very few use cases where this client is actually used.

> Isn't it enough to handle only untyped (*unstructured.Unstructured) and typed (*topolvmv1.LogicalVolume) cases?

Since `wrappedClient` may be used everywhere in the topolvm, I think it would be safer to support `*metav1.PartialObjectMetadata` as well.

Reply:
Thank you for your explanation. I understand.